src/data/equiv/basic.lean

```diff
@@ -1740,6 +1740,31 @@ lemma of_bijective_apply_symm_apply {α β} (f : α → β) (hf : bijective f) (
   (of_bijective f hf).symm (f x) = x :=
 (of_bijective f hf).symm_apply_apply x
 
+section
+
+variables {α' β' : Type*} (e : perm α') {f' : α' → β'}
+  (f : α' ≃ _root_.set.range f') [decidable_pred (∈ _root_.set.range f')]
+
+/--
+Extend the domain of `e : equiv.perm α`, mapping it through `f : α ≃ set.range f'`, for some
+`f' : α → β`. Everything outside of `set.range f` is kept fixed. The `equiv` given by `f` can
+be constructed by `equiv.of_injective` or when there is a known inverse,
+`equiv.set.range_of_left_inverse'`.
+-/
+def perm.via_set_range : perm β' :=
+(perm_congr f e).subtype_congr (equiv.refl _)
+
+@[simp] lemma perm.via_set_range_apply_image (a : α') :
+  e.via_set_range f (f a) = f (e a) :=
```

Comment:
Should we instead have `(f' a)` here? Chances are any real `f` will have a simp lemma taking it to `f'`

Reply:
One isn't promised that the equiv `f` is the bijection with `f'` in the `α' -> set.range f'` direction.

Reply:
Hmm, then isn't `set.range f'` a distraction, and this generalizes to:
```
variables {α' β' : Type*} (e : perm α') {p : β' → Prop}
  (f : α' ≃ subtype p) [decidable_pred p]

def perm.via_subtype : perm β' :=
(perm_congr f e).subtype_congr (equiv.refl _)
```

Reply:
What does that generalization give us? Would I be defining `perm.extend_domain` anyway?

Reply:
It removes the irrelevant detail that `p = (∈ set.range f')` - none of your proofs or definitions use it.